### PR TITLE
Draft: Empty branch

### PR DESCRIPTION
**Pull request by nicolas-vgl-mf**
Empty branch for review

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,180 +0,0 @@
-# Azure ASC Pulumi Dynamic Provider (TS)
```
Comment (nicolas-vgl-mf) [inline README.md:1]: Replacing ASC with  App Service Certificate wil  allow to understand at one glance what this projet is about.
No need to add TS since it is mentionned in the next line

Proposal :  

A Pulumi Dynamic Provider for Azure App Service Certificate

File: README.md [hunk 2]
```diff
@@ -1,180 +0,0 @@
...
-
-The goal of this library is to provide a simple way to provision and maintain ✨App Service Certificat✨, using Typescript/Javascript, with Pulumi.
```
Comment (nicolas-vgl-mf) [inline README.md:3]: Typo :  ✨App Service Certificat✨ =>  ✨App Service Certificate✨

File: README.md [hunk 3]
```diff
@@ -1,180 +0,0 @@
...
-
-The reason why i create this provider is :
```
Comment (nicolas-vgl-mf) [inline README.md:5]: Typo i => I

File: README.md [hunk 4]
```diff
@@ -1,180 +0,0 @@
...
-
--   with Azure Classic provider, is not possible to create an ASC (App Service Provider)
```
Comment (nicolas-vgl-mf) [inline README.md:7]: With Azure Classic it's possible but not what we want in a real use case. I propose something like that : 

```yaml
with Azure Classic provider, it's not possible to create easily an ASC since the module is extremely limited in functionality
```

File: examples/src/kv.ts [hunk 5]
```diff
@@ -1,85 +0,0 @@
...
-// Not sure mandatory, App Service can get a cert into this kv
-const azureAppServiceRegistration = azuread.getServicePrincipal({
-    displayName: 'Microsoft Azure App Service',
-})
-
-const userObjectId = azuread.getUser({
-    userPrincipalName: userName,
-})
-
-const kvname = `kv-test-asc-foobar`
```
Comment (nicolas-vgl-mf) [inline examples/src/kv.ts:30]: Here we need to randomize the keyvault name to prevent Already Exist error when redeploying, something like that


```bash
npm i @pulumi/random
```

```typescript

import * as random from '@pulumi/random'
// .....
const randomSuffix = new random.RandomString(`kv-rnd-name`, {
    length: 3,
    upper: false,
    special: false,
})

const kvName = pulumi.interpolate`kv-test-asc-${randomSuffix.result}`

export const kv = new keyvault.Vault(
    'kv-test',
```

File: package.json [hunk 6]
```diff
@@ -1,64 +0,0 @@
...
-        "App Service",
-        "Certificate"
-    ],
-    "files": [
-        "lib/**/*"
-    ],
-    "main": "lib/index.js",
-    "types": "lib/index.d.ts",
-    "scripts": {
-        "postinstall": "npx enforce-node-version",
```
Comment (nicolas-vgl-mf) [inline package.json:26]: This script requires to install **enforce-node-version** as a Dev dependencies

File: package.json [hunk 7]
```diff
@@ -1,64 +0,0 @@
...
-        "build": "rm -rf ./lib && tsc -p tsconfig-build.json",
-        "format": "prettier --write .",
-        "format-check": "prettier --check .",
-        "lint": "eslint . --ext .ts",
-        "audit-check": "npm audit --registry=https://registry.npmjs.org --audit-level=high",
-        "prepare": "npm run build",
-        "dep-check": "depcheck --ignores=@types/node",
```
Comment (nicolas-vgl-mf) [inline package.json:33]: depcheck is not installed in dev dependencies, you can remove this line or add the package

File: src/ascCertificate.ts [hunk 8]
```diff
@@ -1,571 +0,0 @@
...
-export interface AscCertificateOrderOutputs extends AscCertificateOrderInputs, AscCertificatePrepareOutputs {
-    domainVerificationToken: string
-}
-
-/**
- * Class AscCertificateOrderProvider (Internal) as Pulumi Dynamic Provider
- *
- * This class isn't into Pulumi stack Context (important to understand)
- *
- * @param subscriptionId - Sub Id where is the Kv and wehre you storge your App Service Certificate
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:73]: Typo: wehre => where

File: src/ascCertificate.ts [hunk 9]
```diff
@@ -1,571 +0,0 @@
...
-        let response: string = fqdn
-
-        //replace * by wildcard and check Cert was a wildcard type
-        if (AscCertificateOrderProvider.isWildcard(response)) {
-            response = AscCertificateOrderProvider.replaceAll(response, '*', 'wildcard')
-        }
-        // Replace all '.' by '-'
-        response = AscCertificateOrderProvider.replaceAll(response, '.', '-')
-
-        if (suffix && suffix != 'undefined') {
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:137]: You can just test if the variable is undefined or not

(suffix && suffix != 'undefined') => (suffix != 'undefined')

File: src/ascCertificate.ts [hunk 10]
```diff
@@ -1,571 +0,0 @@
...
-    }
-
-    /**
-     * Call when you make a pulumi refresh
-     * @param id - unique name
-     * @param props - AscCertificateOrderInputs Interface
-     * @returns AscCertificateOrderOutputs Interface
-     */
-    async read(id: string, props: AscCertificateOrderInputs): Promise<pulumi.dynamic.ReadResult> {
-        // pulumi.log.info('Read Methode')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:200]: You can remove this comment

File: src/ascCertificate.ts [hunk 11]
```diff
@@ -1,571 +0,0 @@
...
-
-        const params = this.prepareOutputs(props)
-        const client = this.clientConnect()
-
-        // Create ASC Certificate
-        // pulumi.log.info('appServiceCertificateOrders.get()')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:206]: You can remove this comment

File: examples/src/certificate.ts [hunk 12]
```diff
@@ -1,24 +0,0 @@
-import * as pulumi from '@pulumi/pulumi'
-// import { CertificateOrder } from '../../src/ascCertificate'
```
Comment (nicolas-vgl-mf) [inline examples/src/certificate.ts:2]: You can remove this comment

File: src/ascCertificate.ts [hunk 13]
```diff
@@ -1,571 +0,0 @@
...
-     * Check properties compliance and Which properties are on failure otherwise, just inputs properties
-     *
-     * @param currentOutputs
-     * @param news
-     */
-    async check(
-        currentOutputs: AscCertificateOrderOutputs,
-        news: AscCertificateOrderOutputs,
-    ): Promise<pulumi.dynamic.CheckResult> {
-        // pulumi.log.info('Check Methode')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:243]: You can remove this comment

File: src/ascCertificate.ts [hunk 14]
```diff
@@ -1,571 +0,0 @@
...
-     * @param id
-     * @param previousOutput
-     * @param news
-     */
-    async diff(
-        id: string,
-        previousOutput: AscCertificateOrderOutputs,
-        news: AscCertificateOrderInputs,
-    ): Promise<pulumi.dynamic.DiffResult> {
-        // pulumi.log.info('Diff Methode')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:293]: You can remove this comment

File: src/ascCertificate.ts [hunk 15]
```diff
@@ -1,571 +0,0 @@
...
-            previousOutput.fqdn !== news.fqdn ||
-            previousOutput.suffix !== news.suffix ||
-            previousOutput.resourceGroupName !== news.resourceGroupName ||
-            previousOutput.keyVaultId !== news.keyVaultId
-        ) {
-            changes = true
-            deleteBeforeReplace = true
-
-            if (previousOutput.fqdn !== news.fqdn) {
-                //trigger
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:309]: You can remove this comment

File: src/ascCertificate.ts [hunk 16]
```diff
@@ -1,571 +0,0 @@
...
-
-            if (previousOutput.resourceGroupName !== news.resourceGroupName) {
-                replaces.push('resourceGroupName')
-            }
-            if (previousOutput.keyVaultId !== news.keyVaultId) {
-                replaces.push('keyVaultId')
-            }
-        }
-
-        // Make an update and not a replace
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:324]: For me the comment means that deleteBeforereplace must be set to false, either change the comment or the code

File: src/ascCertificate.ts [hunk 17]
```diff
@@ -1,571 +0,0 @@
...
-     * @param olds
-     * @param news
-     * @returns AscCertificateOrderOutputs Interface
-     */
-    async update(
-        id: string,
-        olds: AscCertificateOrderOutputs,
-        news: AscCertificateOrderInputs,
-    ): Promise<pulumi.dynamic.UpdateResult> {
-        // pulumi.log.info('Update Methode')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:396]: You can remove this comment

File: src/ascCertificate.ts [hunk 18]
```diff
@@ -1,571 +0,0 @@
...
-        return { outs: outs }
-    }
-
-    /**
-     * Call when you remove your ressources from your IAC or 'Pulumi destroy'
-     * @param id
-     * @param props
-     */
-    async delete(id: string, props: AscCertificateOrderInputs) {
-        // pulumi.log.info(`Delete Methode - id: ${id}`)
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:431]: You can remove this comment

File: src/ascCertificate.ts [hunk 19]
```diff
@@ -1,571 +0,0 @@
...
-            params.certificateOrderName,
-            this.linkToKvName,
-            {
-                keyVaultId: inputs.keyVaultId,
-                keyVaultSecretName: params.keyVaultSecretName,
-                location: this.location,
-            },
-        )
-
-        //build formated fqdn how wil be use for other properties
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:371]: I don't undestand, could you rephrase it or remove this comment

File: src/ascCertificate.ts [hunk 20]
```diff
@@ -1,571 +0,0 @@
...
-        olds: AscCertificateOrderOutputs,
-        news: AscCertificateOrderInputs,
-    ): Promise<pulumi.dynamic.UpdateResult> {
-        // pulumi.log.info('Update Methode')
-
-        // updated only if autorenew change
-        const params = this.prepareOutputs(news)
-        const client = this.clientConnect()
-
-        // Create ASC Certificate
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:402]: Update ... You can remove this comment, with the following line it's clear that we're making an update

File: src/ascCertificate.ts [hunk 21]
```diff
@@ -1,571 +0,0 @@
...
-        this.certificateOrderName = asc.certificateOrderName
-        this.certificateURI = asc.certificateURI
-        this.domainVerificationToken = asc.domainVerificationToken
-        this.fqdn = asc.fqdn
-        this.keyVaultId = asc.keyVaultId
-        this.keyVaultSecretName = asc.keyVaultSecretName
-        this.resourceGroupName = asc.resourceGroupName
-        this.suffix = asc.suffix
-    }
-}
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:571]: The file is a too long, to improve lisibility you can split it in 2 or 3

File: src/ascCertificate.ts [hunk 22]
```diff
@@ -1,571 +0,0 @@
...
-
-        //replace * by wildcard and check Cert was a wildcard type
-        if (AscCertificateOrderProvider.isWildcard(response)) {
-            response = AscCertificateOrderProvider.replaceAll(response, '*', 'wildcard')
-        }
-        // Replace all '.' by '-'
-        response = AscCertificateOrderProvider.replaceAll(response, '.', '-')
-
-        if (suffix && suffix != 'undefined') {
-            // this condition test when suffix is converted into a string by is undefined. if that, the suffix is set with value 'undefined'
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:138]: For me this comment is not needed, and less clear than the code itself :)

File: src/ascCertificate.ts [hunk 23]
```diff
@@ -1,571 +0,0 @@
...
-        }
-    }
-
-    /**
-     * Call when 'pulumi up'
-     * @param inputs
-     * @returns AscCertificateOrderOutputs Interface
-     */
-    async create(inputs: AscCertificateOrderInputs): Promise<pulumi.dynamic.CreateResult> {
-        // pulumi.log.info('Create Methode')
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:342]: You can remove this comment

File: src/ascCertificate.ts [hunk 24]
```diff
@@ -1,571 +0,0 @@
...
-                certificateURI: undefined,
-                domainVerificationToken: undefined,
-            },
-            opts,
-        )
-    }
-}
-
-/**
- * Component CertificateOrder
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:515]: Component CertificateOrder => CertificateOrder Component

File: src/ascCertificate.ts [hunk 25]
```diff
@@ -1,571 +0,0 @@
...
-    public readonly autoRenew: pulumi.Output<boolean>
-    public readonly suffix: pulumi.Output<string>
-    public readonly certificateOrderName: pulumi.Output<string>
-    public readonly keyVaultSecretName: pulumi.Output<string>
-    public readonly certificateURI: pulumi.Output<string>
-    public readonly domainVerificationToken: pulumi.Output<string>
-
-    constructor(args: AscCertificateOrderRessourceInputs, opts?: pulumi.CustomResourceOptions) {
-        const commonName = AscCertificateOrderProvider.normalizeFqdn(`${args.fqdn}`, `${args.suffix}`)
-        //<package>:<module>:<type>
```
Comment (nicolas-vgl-mf) [inline src/ascCertificate.ts:554]: You can remove this comment

File: examples/src/kv.ts [hunk 26]
```diff
@@ -1,85 +0,0 @@
...
-                    objectId: azureAppServiceRegistration.then((sp) => sp.id),
-                    permissions: {
-                        secrets: [keyvaultEnums.SecretPermissions.Get],
-                        certificates: [keyvaultEnums.CertificatePermissions.Get],
-                    },
-                    tenantId: tenantId,
-                },
-                {
-                    objectId: userObjectId.then((o) => o.id), // Your User Object ID
-                    // objectId: '10a78cce-ec22-4726-8478-4eb1906192a5', // My User Object ID
```
Comment (nicolas-vgl-mf) [inline examples/src/kv.ts:57]: You can remove this comment

File: examples/.prettierrc [hunk 27]
```diff
@@ -1,15 +0,0 @@
-{
```
Comment (nicolas-vgl-mf) [inline examples/.prettierrc:1]: You don't really need prettier/prettierignore ,you can use  the files defined in the project root directory, in that case you can remove the  dev packages and scripts in package.json

Same thing for eslint

File: examples/.gitignore [hunk 28]
```diff
@@ -1,6 +0,0 @@
-/bin/
-/node_modules/
```
Comment (nicolas-vgl-mf) [inline examples/.gitignore:2]: You don't really need that file, in the root gitignore file use

**node_modules/** instead of **/node_modules/**

File: examples/README.md [hunk 29]
```diff
@@ -1,38 +0,0 @@
-# Pulumi Example
-
-## Pulumi.[stack].yml
-
-In this file, you must specify
-
-Azure Native :
```
Comment (nicolas-vgl-mf) [inline examples/README.md:7]: To make it more clear (it gives the impression that we must configure the 2 sections), I propose to change : 

 `Azure Native` =>` If you're using Azure Native`
 `Azure Classic` =>` If you're using Azure Classic`

File: examples/README.md [hunk 30]
```diff
@@ -1,38 +0,0 @@
...
-  azure:environment: public
-  azure:location: <location>
-  azure:subscriptionId: <xxxx-xxxx>
-  azure:tenantId: <xxxx-xxxx>
-```
-
-set your azure login name
-
-```yaml
-user:name: name@domain.tld
```
Comment (nicolas-vgl-mf) [inline examples/README.md:30]: May be explain why we need that

File: examples/src/kv.ts [hunk 31]
```diff
@@ -1,85 +0,0 @@
-import * as keyvault from '@pulumi/azure-native/keyvault'
-import * as resources from '@pulumi/azure-native/resources'
-import { keyvault as keyvaultEnums } from '@pulumi/azure-native/types/enums'
-import * as azuread from '@pulumi/azuread'
-import * as pulumi from '@pulumi/pulumi'
-
-export const tenantId = new pulumi.Config('azure-native').require('tenantId')
-const userName = new pulumi.Config('user').require('name')
-
-// Common ressource group
```
Comment (nicolas-vgl-mf) [inline examples/src/kv.ts:10]: You can remove this comment

File: examples/src/kv.ts [hunk 32]
```diff
@@ -1,85 +0,0 @@
...
-export const rg = new resources.ResourceGroup(rgName, {
-    resourceGroupName: rgName,
-})
-
-//MANDATORY DO NOT CHANGE
-// ADD Microsoft.Azure.CertificateRegistration acces to kv
-const certificateRegistration = azuread.getServicePrincipal({
-    displayName: 'Microsoft.Azure.CertificateRegistration',
-})
-// Not sure mandatory, App Service can get a cert into this kv
```
Comment (nicolas-vgl-mf) [inline examples/src/kv.ts:21]: If this is not required for the example, you can remove it

File: examples/src/kv.ts [hunk 33]
```diff
@@ -1,85 +0,0 @@
...
-                        secrets: [keyvaultEnums.SecretPermissions.All],
-                    },
-                    tenantId: tenantId,
-                },
-            ],
-            enabledForDeployment: true,
-            enabledForDiskEncryption: true,
-            enabledForTemplateDeployment: true,
-            enableSoftDelete: true,
-            // enablePurgeProtection: false,
```
Comment (nicolas-vgl-mf) [inline examples/src/kv.ts:70]: You can remove this comment

**Review by nicolas-vgl-mf — COMMENTED** [review]
see also the comment on cpa-2303

File: examples/src/kv.ts (hunk 5 above)
Comment (stawen) [reply examples/src/kv.ts:30]: yes I know, i had this in the beginning. But you right, all user can't have the same kv name

**Comment by stawen** [comment]
Thanks @nicolas-vgl-mf for your review. Now I will be able to publish version 1.0.0 of the component.
